python/revng/internal/_cpp_pypeline.cpp

```diff
@@ -0,0 +1,91 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "nanobind/nanobind.h"
+#include "nanobind/stl/optional.h"
+#include "nanobind/stl/set.h"
+#include "nanobind/stl/string.h"
+
+#include "revng/Pypeline/Helpers/Python/Casters.h"
+#include "revng/Pypeline/Helpers/Python/Registry.h"
+#include "revng/Pypeline/Model.h"
+#include "revng/Support/InitRevng.h"
+
+NB_MODULE(_cpp_pypeline, m) {
+  {
+    // Do InitRevng, use a capsule to call the destructor when the Python module
+    // is unloaded
+    int Argc = 1;
+    const char *Argv[] = { "" };
+    const char **ArgvPtr = Argv;
+    nanobind::capsule Capsule(new revng::InitRevng(Argc, ArgvPtr, "", {}),
+                              [](void *Ptr) noexcept {
+                                delete static_cast<revng::InitRevng *>(Ptr);
+                              });
+    nanobind::setattr(m, "__init_revng__", Capsule);
+  }
+
+  using namespace revng::pypeline::helpers::python;
+  nanobind::module_ BaseModule = nanobind::module_::import_("revng.pypeline");
+
+  // Register the Buffer class
+  nanobind::class_<revng::pypeline::Buffer>(m, "Buffer")
+    .def(nanobind::init<>())
+    .def("__buffer__", [](revng::pypeline::Buffer &Handle, int Flags) {
+      return nanobind::steal(PyMemoryView_FromMemory(Handle.data().data(),
+                                                     Handle.data().size(),
+                                                     Flags));
```

Comment:
Maybe `Buffer` should have a method doing `PyMemoryView_FromMemory`. Better data hiding.

Reply:
`Buffer` is exposed to the rest of the pipeline, having it calling python methods would be problematic

Reply:
I got confused with `ManagedPyBuffer`. Let me review this again.

Reply:
Add a comment about the fact that `PyMemoryView_FromMemory` returns a new reference and therefore we need to use `steal` as opposed to `borrow`.